worker/dbaccessor/tracker_test.go

```diff
@@ -0,0 +1,309 @@
+	s.dbApp.EXPECT().Open(gomock.Any(), "controller").Return(s.DB(), nil).Times(DefaultVerifyAttempts - 1)
+	s.dbApp.EXPECT().Open(gomock.Any(), "controller").Return(s.DB(), nil).DoAndReturn(func(_ context.Context, _ string) (*sql.DB, error) {
+		defer close(dbChange)
+		return s.DB(), nil
+	})
+
+	var count uint64
+	verifyFn := func(context.Context, *sql.DB) error {
+		val := atomic.AddUint64(&count, 1)
+
+		if val == DefaultVerifyAttempts {
+			return nil
+		}
+		return errors.New("boom")
+	}
+
+	w, err := NewTrackedDBWorker(s.dbApp, "controller", WithClock(s.clock), WithLogger(s.logger), WithVerifyDBFunc(verifyFn))
+	c.Assert(err, jc.ErrorIsNil)
+
+	defer workertest.DirtyKill(c, w)
+
+	select {
+	case <-done:
+	case <-time.After(testing.ShortWait):
+		c.Fatal("timed out waiting for DB callback")
+	}
+
+	// The db should have changed to the new db.
+	select {
+	case <-dbChange:
+	case <-time.After(testing.ShortWait):
+		c.Fatal("timed out waiting for DB callback")
+	}
+
+	tables := checkTableNames(c, w)
+	c.Assert(tables, SliceContains, "lease")
+
+	workertest.CleanKill(c, w)
+
+	c.Assert(w.Err(), jc.ErrorIsNil)
+}
+
+func (s *trackedDBWorkerSuite) TestWorkerAttemptsToVerifyDBButSucceedsWithDifferentDB(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.expectAnyLogs()
+	s.expectClock()
+	done := s.expectTimer(1)
+
+	s.logger.EXPECT().Errorf(gomock.Any(), gomock.Any()).AnyTimes()
+
+	dbChange := make(chan struct{})
+	exp := s.dbApp.EXPECT()
+	gomock.InOrder(
+		exp.Open(gomock.Any(), "controller").Return(s.DB(), nil),
+		exp.Open(gomock.Any(), "controller").Return(s.DB(), nil),
+		exp.Open(gomock.Any(), "controller").DoAndReturn(func(_ context.Context, _ string) (*sql.DB, error) {
+			defer close(dbChange)
+			return s.NewCleanDB(c), nil
+		}),
+	)
+
+	var count uint64
+	verifyFn := func(context.Context, *sql.DB) error {
+		val := atomic.AddUint64(&count, 1)
+
+		if val == DefaultVerifyAttempts {
+			return nil
+		}
+		return errors.New("boom")
+	}
+
+	w, err := NewTrackedDBWorker(s.dbApp, "controller", WithClock(s.clock), WithLogger(s.logger), WithVerifyDBFunc(verifyFn))
+	c.Assert(err, jc.ErrorIsNil)
+
+	defer workertest.DirtyKill(c, w)
+
+	select {
+	case <-done:
+	case <-time.After(testing.ShortWait):
+		c.Fatal("timed out waiting for DB callback")
+	}
+
+	// The db should have changed to the new db.
+	select {
+	case <-dbChange:
+	case <-time.After(testing.ShortWait):
+		c.Fatal("timed out waiting for DB callback")
+	}
+
+	tables := checkTableNames(c, w)
+	c.Assert(tables, gc.Not(SliceContains), "lease")
+
+	workertest.CleanKill(c, w)
+
+	c.Assert(w.Err(), jc.ErrorIsNil)
+}
+
+func (s *trackedDBWorkerSuite) TestWorkerAttemptsToVerifyDBButFails(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.expectAnyLogs()
+	s.expectClock()
+	done := s.expectTimer(1)
+
+	s.logger.EXPECT().Errorf(gomock.Any(), gomock.Any()).AnyTimes()
+
+	s.dbApp.EXPECT().Open(gomock.Any(), "controller").Return(s.DB(), nil).Times(DefaultVerifyAttempts)
```

Comment:
This checks that we only check 3 times to retry connecting when a ping fails,
but it doesn't check that if it succeeds on the first verify, that we later do a follow up verify 10s later.